README.md
# pm2-hooks

[![GitHub version][fury-badge]][fury-url]
[![Travis-CI][travis-badge]][travis-url]
[![Codeclimate][codeclimate-badge]][codeclimate-url]
[![Codeclimate Coverage][codeclimate-cov-badge]][codeclimate-cov-url]
[![Dependency][david-badge]][david-url]
[![DevDependency][david-dev-badge]][david-dev-url]

PM2 module to listen webhooks from github, bitbucket, gitlab, jenkins and droneci. When a webhook is received you can run a script, pull your project, restart pm2, etc.

This project is highly inspired by [vmarchaud/pm2-githook](https://github.com/vmarchaud/pm2-githook).

Features ([changelog](/CHANGELOG.md)):

- Runs an http server listening for webhooks
- Works on any repository system, as for now it does nothing with the payload received. In the near future I will check the branch or the action, the secret, etc.
- Only runs the command you set
- Run the command in the cwd defined for the app

Wanted features, to be done during Mach/2017:

- Check payload for secret, check common headers on main git repositories (github, bitbucket, gitlab, etc) to know if is a valid call

Possible features, as I need to think about it:

- Auto-restart pm2 app after a successful command run
- Make an automatic `git pull` on the folder, and make a `prePull` and `postPull` available commands (same approach as [vmarchaud/pm2-githook](https://github.com/vmarchaud/pm2-githook))

# Install

To install it simply run:

```bash
$ pm2 install pm2-hooks
```

Warning: This library currently (2017 feb 26) is in ALPHA state. This means some things:

- You can help me a lot making a comment/issue/PR
- I will try to publish the last version to npm so you can install it with only `pm2 install pm2-hooks`. If for some reason the version on npm is outdated you always will be capable of run `pm2 install desaroger/pm2-hooks` to be sure to install the last version directly from the repository.


# Usage

## Step 1: Prepare the ecosystem file

By default **pm2-hooks** doesn't do anything. You need to set the key `env_hook` inside the config of a given app, inside the ecosystem file.

If env_hook isn't defined or is falsy then is disabled.

Example of an ecosystem file:

```js
{
	apps: [
		{
			name: 'api-1',
			script: 'server.js'
		},
		{
			name: 'api-2',
			script: 'server.js',
			env_hook: {
				command: 'git pull && npm i && npm test && pm2 restart api-2',
				cwd: '/home/desaroger'
			}
		}
	]
}
```

Where **api-1** has hook disabled and **api-2** is enabled and when the hook is called, the command is executed.

### Available options:

- **command** *{string}* The line you want to execute. Will be executed with NodeJS `spawn`. (optional, but if not set this is not going to do nothing ¯\\_(ツ)_/¯)
- **cwd** *{string}* The cwd to use when running the command. If not set, the one used on your ecosystem app configuration will be used (if set).
- **commandOptions** *{object}* The object that we will pass to the NodeJS `spawn`. Defaults to a blank object, and later we add the *cwd*.
- **type** *{string}* [not implemented yet] The git server you are going to use [github, gitlab, bitbucket, etc].

## Step 2: Install

If you didn't install before, install it. If you installed it, then you will need to restart it. For that, run `pm2 restart pm2-hooks`.

## Step 3: Try it

Now you have a server on port 9000 by default. You can make a call to `http://localhost:9000/api-2` to see the response.

If everything went fine, you will see:

```js
{
	status: 'success',
	message: 'Route "api-2" was found'
	code: 0
}
```

And the command had been executed.

## Step 4: See the log

Everything will be logged in the pm2 logs. For see them, run:

```bash
$ pm2 logs pm2-hooks
```

And for see the entire log:

```bash
$ cat ~/.pm2/logs/pm2-hooks-out-0.log
```

# FAQ

## How I can change the port?

You can set the port (where the default port is 9000) setting it in the config of the pm2 module. For doing that, run:

```bash
$ pm2 set pm2-hooks:port 3000
```

## How I can uninstall it?

You can uninstall this module running:

```bash
$ pm2 uninstall pm2-hooks
```

# Another similar projects

These are some projects I found similar to mine. Please let me know if you know anoher.

- [vmarchaud/pm2-githook](https://github.com/vmarchaud/pm2-githook): From where I was inspired. It works on any repository, pulls the repo when webhook is called and has *preHook* and *postHook* commands available.
- [oowl/pm2-webhook](https://github.com/oowl/pm2-webhook): Works on any repository. If you want to use the *secret*, then the webhook must contain the *X-Hub-Signature* in order to work (I don't know if every git server contains it).

# Copyright and license

Copyright 2017 Roger Fos Soler

Licensed under the [MIT License](/LICENSE).


[npm-badge]: https://img.shields.io/npm/v/pm2-hooks.svg
[npm-url]: https://www.npmjs.com/package/pm2-hooks

[fury-badge]: https://badge.fury.io/js/pm2-hooks.svg
[fury-url]: https://www.npmjs.com/package/pm2-hooks

[travis-badge]: https://travis-ci.org/desaroger/pm2-hooks.svg
[travis-url]: https://travis-ci.org/desaroger/pm2-hooks

[david-badge]: https://david-dm.org/desaroger/pm2-hooks.svg
[david-url]: https://david-dm.org/desaroger/pm2-hooks
[david-dev-badge]: https://david-dm.org/desaroger/pm2-hooks/dev-status.svg
[david-dev-url]: https://david-dm.org/desaroger/pm2-hooks#info=devDependencies

[gemnasium-badge]: https://gemnasium.com/badges/github.com/desaroger/pm2-hooks.svg
[gemnasium-url]: https://gemnasium.com/github.com/desaroger/pm2-hooks

[codeclimate-badge]: https://codeclimate.com/github/desaroger/pm2-hooks/badges/gpa.svg
[codeclimate-url]: https://codeclimate.com/github/desaroger/pm2-hooks

[codeclimate-cov-badge]: https://codeclimate.com/github/desaroger/pm2-hooks/badges/coverage.svg?hash=1
[codeclimate-cov-url]: https://codeclimate.com/github/desaroger/pm2-hooks/coverage

[coverage-badge]: https://codeclimate.com/github/desaroger/pm2-hooks/badges/coverage.svg
[coverage-url]: https://codeclimate.com/github/desaroger/pm2-hooks/coverage


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## 1.1.13/14 - 2019-07-12
### Fix
- Respond to the webhook server without waiting the command to end. Fixes issues with Github Webhook timeout(10s).

## 1.1.12 - 2019-07-06
### Fix
- Hooks with secret not working as not being able to hash the raw body

## 1.1.10/11 - 2019-07-06
### Fix
- Issue with bitbucket body not being parsed correctly

### Change
- Now travis test on multiple node versions

## 1.1.9 - 2018-08-05
### Fix
- Issue with github webhook without secret, not receiving the header

## 1.1.5-8 - 2018-06-01
### Change
- Upgrade dependencies

### Fix
- Readme typo

## 1.1.4 - 2017-03-07
### Add
- Initial support for bitbucket

## 1.1.2-1.1.3 - 2017-03-04
### Add
- Initial secret support for github

## 1.1.0-1.1.1 - 2017-03-04
### Add
- CWD configuration support

## 1.0.7 - 2017-02-26
### Add
- Badges

## 1.0.6 - 2017-02-26
### Add
- Babel
- Readme

## 1.0.5 - 2017-02-26
### Add
- Log service

## 1.0.4 - 2017-02-26
### Add
- Run of a simple command

## 1.0.3 - 2017-02-26
### Add
- Index on root folder

## 1.0.2 - 2017-02-25
### Add
- Basics on Pm2Module
- More tests to WebhookServer

## 1.0.1 - 2017-02-23
### Add
- Test and lint

## 1.0.0 - 2017-02-23
### Add
- Base of the project


test/utils.js
/**
 * Created by desaroger on 7/03/17.
 */

let { expect, isPromise } = require('./assets');

describe('utils', () => {
    describe('[isPromise]', () => {
        it('is a function', () => {
            expect(isPromise).to.be.a('function');
        });

        it('returns false if falsy', () => {
            expect(isPromise()).to.equal(false);
            expect(isPromise(false)).to.equal(false);
            expect(isPromise(null)).to.equal(false);
        });

        it('returns false if non-object', () => {
            expect(isPromise('nope')).to.equal(false);
        });

        it('returns false if .then is not a function', () => {
            expect(isPromise({})).to.equal(false);
            expect(isPromise({ then: 'hi' })).to.equal(false);
        });

        it('returns false if .catch is not a function', () => {
            expect(isPromise({})).to.equal(false);
            expect(isPromise({ then: 'hi' })).to.equal(false);
            expect(isPromise({ then: 'hi', catch: 'hi' })).to.equal(false);
        });

        it('returns true if another case', () => {
            let promise = {
                then() {},
                catch() {}
            };
            expect(isPromise(promise)).to.equal(true);
        });
    });
});


test/assets/c.js
/**
 * Created by desaroger on 23/02/17.
 */

let co = require('co');

let c = co.wrap;
c.run = co;

module.exports = c;


test/Pm2Module.js
/**
 * Created by desaroger on 25/02/17.
 */

let _ = require('lodash');
let Pm2Module = require('../src/Pm2Module');
let WebhookServer = require('../src/WebhookServer');
let mockApps = require('./mocks/apps');
let { expect, c, callApi, mockSpawn, log } = require('./assets');

describe('Pm2Module', () => {
    afterEach(() => {
        log.restore();
    });

    it('is a function', () => {
        expect(Pm2Module).to.be.a('function');
    });

    it('can be instantiated', () => {
        expect(() => new Pm2Module()).to.not.throw();
    });

    describe('instance', (pm2Module) => {
        before(() => {
            let apps = [
                wrapEnv({ name: 'a' }, { type: 'github' }),
                wrapEnv({ name: 'b' }, { type: 'bitbucket' }),
                wrapEnv({ name: 'c' }, { type: 'gitlab' }),
                wrapEnv({ name: 'd' }, { })
            ];
            pm2Module = new Pm2Module(apps);
        });

        it('has the method start', () => {
            expect(pm2Module.start).to.be.a('function');
        });

        it('has the method stop', () => {
            expect(pm2Module.stop).to.be.a('function');
        });

        it('has created the server, but not started', () => {
            expect(pm2Module.webhookServer).to.be.an.instanceOf(WebhookServer);
            expect(pm2Module.webhookServer.isRunning()).to.equal(false);
        });

        it('the server has the routes created', () => {
            let routes = pm2Module.webhookServer.options.routes;
            expect(routes.a).to.shallowDeepEqual({ type: 'github' });
            expect(routes.b).to.shallowDeepEqual({ type: 'bitbucket' });
            expect(routes.c).to.shallowDeepEqual({ type: 'gitlab' });
            expect(routes.d).to.be.an('object');
            expect(routes.d.type).to.not.be.ok;
        });
    });

    describe('[run commands]', (pm2Module) => {
        before(() => {
            let apps = [
                wrapEnv({
                    name: 'a'
                }, {
                    command: 'echo hi',
                    cwd: '/home/desaroger'
                }),
                wrapEnv({
                    name: 'b',
                    pm_cwd: '/home/lol'
                }, {
                    command: 'echo-nope hi'
                }),
                wrapEnv({
                    name: 'c',
                    pm_cwd: '/home/nope'
                }, {
                    command: 'echo yeah',
                    cwd: '/home/yeah'
                }),
                wrapEnv({
                    name: 'd',
                    pm_cwd: '/home/nope'
                }, {})
            ];
            pm2Module = new Pm2Module(apps, {
                port: 1234
            });
            return pm2Module.start();
        });
        after(() => pm2Module.stop());
        afterEach(() => mockSpawn.restore());

        it('is running', () => {
            expect(pm2Module.webhookServer.isRunning()).to.equal(true);
        });

        it('returns the warning if not found', c(function* () {
            log.mock((msg, status) => {
                expect(msg).to.match(/route "nope" not found/i);
                expect(status).to.equal(1);
            });
            let result = yield callApi('/nope');
            expect(result).to.deep.equal({
                $statusCode: 400,
                status: 'warning',
                message: 'Route "nope" not found',
                code: 1
            });
            expect(log.count).to.equal(1);
            log.checkMocks();
        }));

        it('runs the command', c(function* () {
            mockSpawn.start((command, options) => {
                expect(command).to.equal('echo hi');
            });
            log.mock((msg, status) => {
                expect(status).to.equal(0);
                expect(msg).to.match(/Parsed payload: \{}/);
            });
            log.mock((msg, status) => {
                expect(msg).to.match(/running command: echo hi/i);
                expect(status).to.equal(0);
            });
            log.mock((msg, status) => {
                expect(msg).to.match(/route "a" was found/i);
                expect(status).to.equal(0);
            });
            let result = yield callApi('/a');
            expect(result).to.deep.equal({
                $statusCode: 200,
                status: 'success',
                message: 'Route "a" was found',
                code: 0
            });
            expect(log.count).to.equal(3);
            log.checkMocks();
        }));

        it('doesn\'t throw if no command', c(function* () {
            log.mock((msg, status) => {
                expect(status).to.equal(0);
                expect(msg).to.match(/Parsed payload: \{}/);
            });
            log.mock((msg, status) => {
                expect(msg).to.match(/route "d" was found/i);
                expect(status).to.equal(0);
            });
            let result = yield callApi('/d');
            expect(result).to.deep.equal({
                $statusCode: 200,
                status: 'success',
                message: 'Route "d" was found',
                code: 0
            });
            expect(log.count).to.equal(2);
            log.checkMocks();
        }));

        it('runs the command in the CWD if available on config', c(function* () {
            mockSpawn.start((command, options) => {
                expect(options).to.be.an('object');
                expect(options.cwd).to.equal('/home/desaroger');
                expect(command).to.equal('echo hi');
            });
            log.mock((msg, status) => {
                expect(status).to.equal(0);
                expect(msg).to.match(/Parsed payload: \{}/);
            });
            log.mock((msg, status) => {
                expect(status).to.equal(0);
                expect(msg).to.match(/Running command: echo hi/);
            });
            log.mock((msg, status) => {
                expect(status).to.equal(0);
                expect(msg).to.match(/Success: Route "a" was found/);
            });
            let result = yield expect(callApi('/a')).to.be.fulfilled;
            expect(result).to.deep.equal({
                $statusCode: 200,
                status: 'success',
                message: 'Route "a" was found',
                code: 0
            });
            log.checkMocks();
        }));

        it('runs the command in the CWD with the app cwd if no present config', c(function* () {
            mockSpawn.start((command, options) => {
                expect(options).to.be.an('object');
                expect(options.cwd).to.equal('/home/lol');
                expect(command).to.equal('echo-nope hi');
            });
            log.mock((msg, status) => {
                expect(status).to.equal(0);
                expect(msg).to.match(/Parsed payload: \{}/);
            });
            log.mock((msg, status) => {
                expect(status).to.equal(0);
                expect(msg).to.match(/Running command: echo-nope hi/);
            });
            log.mock((msg, status) => {
                expect(status).to.equal(0);
                expect(msg).to.match(/Success: Route "b" was found/);
            });
            let result = yield expect(callApi('/b')).to.be.fulfilled;
            expect(result).to.deep.equal({
                $statusCode: 200,
                status: 'success',
                message: 'Route "b" was found',
                code: 0
            });
            log.checkMocks();
        }));

        it('runs the command in the CWD with priority to the config', c(function* () {
            mockSpawn.start((command, options) => {
                expect(options).to.be.an('object');
                expect(options.cwd).to.equal('/home/yeah');
                expect(command).to.equal('echo yeah');
            });
            log.mock((msg, status) => {
                expect(status).to.equal(0);
                expect(msg).to.match(/Parsed payload: \{}/);
            });
            log.mock((msg, status) => {
                expect(status).to.equal(0);
                expect(msg).to.match(/Running command: echo yeah/);
            });
            log.mock((msg, status) => {
                expect(status).to.equal(0);
                expect(msg).to.match(/Success: Route "c" was found/);
            });
            let result = yield expect(callApi('/c')).to.be.fulfilled;
            expect(result).to.deep.equal({
                $statusCode: 200,
                status: 'success',
                message: 'Route "c" was found',
                code: 0
            });
            log.checkMocks();
        }));

        it('shows error if command error', c(function* () {
            mockSpawn.start(() => {
                return function (cb) {
                    this.emit('close', 'asd');
                    return cb(2);
                };
            });
            log.mock((msg, status) => {
                expect(status).to.equal(0);
                expect(msg).to.match(/Parsed payload: \{}/);
            });
            log.mock((msg, status) => {
                expect(status).to.equal(0);
                expect(msg).to.match(/Running command: echo hi/);
            });
            log.mock((msg, status) => {
                expect(status).to.equal(0);
                expect(msg).to.match(/Route "a" was found/);
            });
            log.mock((msg, status) => {
                expect(status).to.equal(2);
                expect(msg).to.match(/Error on "a" route: asd/);
            });
            let result = yield callApi('/a');
            expect(result).to.deep.equal({
                $statusCode: 200,
                status: 'success',
                message: 'Route "a" was found',
                code: 0
            });
            log.checkMocks();
            expect(log.count).to.equal(4);
        }));
    });

    describe('method _parseProcesses', () => {
        it('exists', () => {
            expect(Pm2Module._parseProcesses).to.be.a('function');
        });

        it('accepts a void array', () => {
            let result = Pm2Module._parseProcesses([]);
            expect(result).to.deep.equal({});
        });

        it('sets "unknown" as default name', () => {
            let result = Pm2Module._parseProcesses([wrapEnv()]);
            expect(Object.keys(result)).to.have.length(1);
            expect(result.unknown).to.be.ok;
        });

        it('works with a real processes array', () => {
            let result = Pm2Module._parseProcesses(mockApps);
            expect(result).to.have.all.keys(['api', 'api2', 'panel']);
            expect(result).to.shallowDeepEqual({
                api: {
                    type: 'bitbucket'
                },
                api2: {},
                panel: {
                    type: 'gitlab'
                }
            });
        });
    });

    describe('method _parseProcess', () => {
        it('exists', () => {
            expect(Pm2Module._parseProcess).to.be.a('function');
        });

        it('returns false if no options specified', () => {
            let objs = {
                a: undefined,
                b: null,
                c: {},
                d: { pm2_env: {} }
            };
            let mockLog = (msg, status) => {
                expect(status).to.equal(0);
                expect(msg).to.match(/No options found for "undefined" route/);
            };
            log.mock(mockLog);

            _.values(objs)
                .forEach((opts) => {
                    expect(Pm2Module._parseProcess(opts)).to.equal(null);
                });
            expect(log.count).to.equal(1);
            log.checkMocks();
        });

        it('returns the route if valid object', () => {
            let obj = wrapEnv({ name: 'lol' }, { type: 'bitbucket' });
            let result = Pm2Module._parseProcess(obj);
            expect(result).to.shallowDeepEqual({
                name: 'lol',
                type: 'bitbucket'
            });
        });

        it('returns the route if equal true', () => {
            let obj = wrapEnv({ name: 'lol' }, true);
            let result = Pm2Module._parseProcess(obj);
            expect(result).to.shallowDeepEqual({
                name: 'lol'
            });
        });
    });

    describe('method _runCommand', () => {
        it('exists', () => {
            expect(Pm2Module._runCommand).to.be.a('function');
        });

        it('returns a promise', () => {
            let result = Pm2Module._runCommand('echo hi');
            expect(result).to.be.ok;
            expect(result.then).to.be.a('function');
            expect(result.catch).to.be.a('function');
        });

        it('runs the line', () => {
            return Pm2Module._runCommand('echo hi')
                .then((result) => {
                    expect(result).to.not.be.ok;
                })
                .catch((err) => {
                    expect(err).to.not.be.ok;
                });
        });

        it('throws an error if fails', () => {
            return Pm2Module._runCommand('765-this-need-to-not-exists-231 hi')
                .catch((err) => {
                    expect(err).to.be.ok;
                });
        });
    });
});

/**
 * Creates an object with the same structure as the PM2 process
 * Example structure:
 * {
 *      name: 'api',
 *      pm2_env: {
 *          env: {},
 *          env_hooks: {
 *              command: 'echo hi'
 *          }
 *      }
 * }
 *
 * @param {object} appData The root object
 * @param {object} envHook The object inside 'pm2_env.env_hooks' path
 * @returns {object} The built object
 */
function wrapEnv(appData = {}, envHook = {}) {
    appData.pm2_env = { env_hook: envHook };
    return appData;
}


src/utils/log.js
/**
 * Created by desaroger on 26/02/17.
 */

function log(msg, status = 0) {
    log.count++;
    msg = `[${log.getDate()}] ${msg}`;
    if (log.mocks.length) {
        let mock = log.mocks.shift();
        if (mock.options.checkNow) {
            mock.method(msg, status);
        } else {
            log.history.push({ method: mock.method, msg, status });
        }
        // return method.call(log, msg, status);
    } else {
        return log.defaultMethod(msg, status);
    }
}

log.count = 0;
log.history = [];

log.getDate = function build() {
    return new Date().toISOString();
};

log.mocks = [];

log.defaultMethod = function defaultMethod(msg, status = 0) {
    // if (process.env.NODE_ENV === 'test') {
    //     // return;
    // }
    let map = {
        0: 'log',
        1: 'warn',
        2: 'error'
    };
    console[map[status]](msg);
};

log.mock = function mock(method = () => {}, options = {}) {
    log.mocks.push({ method, options });
    log.mockedMethod = method;
};

log.restore = function restore() {
    log.count = 0;
    log.mockedOnce = false;
    log.mockedMethod = false;
    log.mocks = [];
    log.history = [];
};

log.checkMocks = function checkMocks() {
    this.history.forEach(({ method, msg, status }) => {
        method(msg, status);
    });
};

module.exports = log;


src/WebhookServer.js
/**
 * Created by desaroger on 23/02/17.
 */

const _ = require('lodash');
const http = require('http');
const bodyParser = require('body-parser');
const crypto = require('crypto');
// const ipaddr = require('ipaddr.js');
const { log, c, isPromise } = require('./utils');

class WebhookServer {

    /**
     * @param {object} options The options
     * @constructor
     */
    constructor(options = {}) {
        _.defaults(options, {
            port: process.env.PORT || 9000,
            routes: {}
        });
        for (let name in options.routes) {
            /* istanbul ignore next */
            if (!options.routes.hasOwnProperty(name)) continue;
            options.routes[name].name = name;
        }
        this.options = options;
    }

    /**
     * Creates the server and start it (with .listen)
     *
     * @returns {Promise<WebhookServer>} Returns the instance
     */
    start() {
        return new Promise((resolve, reject) => {
            if (this.server) {
                return reject('Server previously started');
            }
            this.server = http.createServer(this._handleCall.bind(this));
            this.server.listen(this.options.port, (err) => {
                /* istanbul ignore next */
                if (err) return reject(err);
                resolve(this);
            });
        });
    }

    /**
     * Stops the server. Doesn't throws an error if fails.
     *
     * @returns {Promise<WebhookServer>} Returns the instance
     */
    stop() {
        return new Promise((resolve) => {
            if (!this.server) {
                return resolve(this);
            }
            this.server.close(() => {
                delete this.server;
                resolve(this);
            });
        });
    }

    /**
     * Parses the body of a request
     *
     * @param {http.IncomingMessage} req The Request of the call
     * @param {http.ServerResponse} res The Response of the call
     * @param {function} fn The function to call after the body has been parsed
     * @private
     */
    _parseBody(req, res, fn) {
        let rawBodyParser = (req2, res2, buf, encoding) => {
            if (buf && buf.length) {
                req2.rawBody = buf.toString(encoding || 'utf8');
                console.log("rawBody", req2.rawBody);
            } else {
                console.log("nope");
            }
        };

        bodyParser.urlencoded({
            extended: true,
            verify: rawBodyParser
        })(req, res, () => {
            bodyParser.json({
                verify: rawBodyParser
            })(req, res, () => {
                bodyParser.raw({
                    type: () => true,
                    verify: rawBodyParser
                })(req, res, fn);
            });
        });
    }

    /**
     * This method is the main function of the http server.
     * Given a request, it finds out the matched route and
     * calls the method of the route.
     *
     * @param {http.IncomingMessage} req The Request of the call
     * @param {http.ServerResponse} res The Response of the call
     * @private
     */
    _handleCall(req, res) {
        let self = this;
        this._parseBody(req, res, c(function* () {
            try {
                // Mock
                let routeName = self._getRouteName(req);
                if (!routeName) {
                    log('No route found on url', 1);
                    res.statusCode = 400;
                    res.end(JSON.stringify({
                        status: 'warning',
                        message: 'No route found on url',
                        code: 1
                    }));
                    return;
                }
                let route = self.options.routes[routeName];
                if (!route) {
                    log(`Warning: Route "${routeName}" not found`, 1);
                    res.statusCode = 400;
                    res.end(JSON.stringify({
                        status: 'warning',
                        message: `Route "${routeName}" not found`,
                        code: 1
                    }));
                    return;
                }

                // Prepare the execution of the method
                let payload = self._parseRequest(req, route);
                let result;
                try {
                    result = route.method(payload);
                    if (isPromise(result)) {
                        result = yield result;
                    }
                } catch (e) {
                    let message = e.message ? e.message : e;
                    log(`Error: Route "${routeName}" method error: ${message}`, 2);
                    res.statusCode = 500;
                    res.end(JSON.stringify({
                        status: 'error',
                        message: `Route "${routeName}" method error: ${message}`,
                        code: 2
                    }));
                    return;
                }

                log(`Success: Route "${routeName}" was found`, 0);
                res.end(JSON.stringify({
                    status: 'success',
                    message: `Route "${routeName}" was found`,
                    code: 0,
                    result
                }));
            } catch (e) {
                log(e.message, 2);
                res.statusCode = 500;
                res.end(JSON.stringify({
                    status: 'error',
                    message: `Unexpected error: ${e.message}`,
                    code: 2
                }));
            }
        }));
    }

    /**
     * Given a request, returns the matched route
     *
     * @param {http.IncomingMessage} req The Request of the call
     * @returns {string|null} The matched route or null if not
     * @private
     */
    _getRouteName(req) {
        let path = trimSlashes(req.url);
        let name = path.split('/')[0];
        name = name.split('?')[0];
        if (!name) {
            return null;
        }

        return name;
    }

    _parseRequest(req, route) {
        route = Object.assign({
            name: '<unknown>',
            type: 'auto',
            secret: false
        }, route);

        req.headers = _.transform(req.headers, (result, val, key) => {
            result[key.toLowerCase()] = val;
        });

        // Auto-type
        if (route.type === 'auto') {
            if (req.headers['x-github-event']) {
                route.type = 'github';
            } else if (req.headers['user-agent'] && /bitbucket/i.test(req.headers['user-agent'])) {
                route.type = 'bitbucket';
            }
            if (route.type === 'auto') {
                route.type = false;
            }
        }

        // Checks
        let result = {};
        if (route.type) {
            let checksMap = {
                github() {
                    let error = false;
                    if (!req.headers['x-github-event'] || (route.secret && !req.headers['x-hub-signature'])) {
                        error = 'Invalid headers';
                    } else if (route.secret) {
                        if (!req.rawBody) {
                            error = 'Secret required and body not found on request';
                        } else {
                            let hash = crypto.createHmac('sha1', route.secret);
                            hash = hash.update(req.rawBody).digest('hex');
                            if (`sha1=${hash}` !== req.headers['x-hub-signature']) {
                                error = 'Invalid secret';
                            }
                        }
                    }
                    return error;
                },
                bitbucket() {
                    let error = false;
                    let requiredHeaders = ['x-event-key', 'x-request-uuid'];
                    if (requiredHeaders.some(k => !req.headers[k])) {
                        error = 'Invalid headers';
                    } else if (route.secret) {
                        error = 'Secret not supported for bitbucket yet';
                    } else {
                        error = 'Invalid body';
                        if (req.body && _.isObjectLike(req.body)) {
                            let action = req.headers['x-event-key'].replace('repo:', '');
                            if (!action) {
                                error = 'Invalid headers';
                            } else if (req.body[action]) {
                                error = false;
                            }
                        }
                    }
                    return error;
                },
                test() {}
            };
            let method = checksMap[route.type];
            if (!method) {
                let error = `Error unknown route type ${route.type}`;
                log(error, 2);
                throw new Error(error);
            }

            let error = method();
            if (error) {
                error = `Error for route type ${route.type}: ${error}`;
                log(error, 2);
                throw new Error(error);
            }

            // Parse
            let body = typeof req.body === 'string' ? JSON.parse(req.body) : req.body;
            body = body || {};
            let parseMap = {
                github() {
                    if (body.payload) {
                        body = body.payload;
                    }
                    if (typeof body === 'string') {
                        body = JSON.parse(body);
                    }
                    result.name = _.get(body, 'repository.name');
                    result.action = req.headers['x-github-event'];
                    result.branch = _.get(body, 'ref', '').replace('refs/heads/', '') || false;
                },
                bitbucket() {
                    result.action = req.headers['x-event-key'].replace('repo:', '');
                    result.name = _.get(body, 'repository.name');
                    result.branch = '<unknown>';
                    let changes = _.get(body, `${result.action}.changes`) || [];
                    let validChange = changes
                        .reduce((total, change) => {
                            total = total.concat([change.old, change.new]);
                            return total;
                        }, [])
                        .filter((change) => {
                            return change && change.type === 'branch';
                        })
                        .find((change) => {
                            return !!change.name;
                        });
                    if (!validChange) {
                        error = `Error for route type ${route.type}: Invalid "changes" key on body`;
                        log(error, 2);
                        throw new Error(error);
                    }
                    result.branch = validChange.name;
                },
                test() {
                    result = body;
                }
            };
            parseMap[route.type]();
        }

        return result;
    }

    /**
     * Checks if the internal server is running.
     *
     * @returns {boolean} True if the server is running
     */
    isRunning() {
        return !!(this.server && this.server.address());
    }

}

module.exports = WebhookServer;

function trimSlashes(s) {
    s = s.trim().replace(/(^\/)|(\/$)/, '');
    return s;
}
